test/markup/cpp/template-complexity.expect.txt

```diff
@@ -1,21 +1,21 @@
 <span class="hljs-keyword">template</span> &lt;<span class="hljs-keyword">class</span> <span class="hljs-title class_">T</span>&gt; <span class="hljs-comment">// comment</span>
-<span class="hljs-function"><span class="hljs-keyword">auto</span> <span class="hljs-title">foo</span><span class="hljs-params">(T x)</span> </span>{ ... };
+<span class="hljs-function"><span class="hljs-type">auto</span> <span class="hljs-title">foo</span><span class="hljs-params">(T x)</span> </span>{ ... };
 
 <span class="hljs-keyword">namespace</span> impl {
     <span class="hljs-keyword">template</span>&lt;<span class="hljs-keyword">typename</span> T&gt;
-    <span class="hljs-keyword">struct</span> <span class="hljs-title class_">is_streamable</span>&lt;T, std::<span class="hljs-keyword">void_t</span>&lt;<span class="hljs-keyword">decltype</span>(std::declval&lt;std::wostream &amp;&gt;() &lt;&lt; std::declval&lt;T&gt;())&gt;&gt; : std::true_type { };
+    <span class="hljs-keyword">struct</span> <span class="hljs-title class_">is_streamable</span>&lt;T, std::<span class="hljs-type">void_t</span>&lt;<span class="hljs-keyword">decltype</span>(std::<span class="hljs-built_in">declval</span>&lt;std::wostream &amp;&gt;() &lt;&lt; std::<span class="hljs-built_in">declval</span>&lt;T&gt;())&gt;&gt; : std::<span class="hljs-type">true_type</span> { };
 }
 
 <span class="hljs-comment">// Disable overload for already valid operands.</span>
-<span class="hljs-keyword">template</span>&lt;<span class="hljs-keyword">class</span> <span class="hljs-title class_">T</span>, class = std::<span class="hljs-keyword">enable_if_t</span>&lt;!impl::is_streamable_v&lt;<span class="hljs-keyword">const</span> T &amp;&gt; &amp;&amp; std::is_convertible_v&lt;<span class="hljs-keyword">const</span> T &amp;, std::wstring_view&gt;&gt;&gt;
+<span class="hljs-keyword">template</span>&lt;<span class="hljs-keyword">class</span> <span class="hljs-title class_">T</span>, <span class="hljs-type">class</span> = std::<span class="hljs-type">enable_if_t</span>&lt;!impl::is_streamable_v&lt;<span class="hljs-keyword">const</span> T &amp;&gt; &amp;&amp; std::is_convertible_v&lt;<span class="hljs-keyword">const</span> T &amp;, std::<span class="hljs-type">wstring_view</span>&gt;&gt;&gt;
```

Comment:
I don't believe class is a type here?  We should probably remove it from the list of types and (in a follow-up PR) add it instead with a `class [IDENT]` multi-matcher rule as done in many other languages - allowing us to also flag the class name as a `title.class`, etc...

Reply:
Also pretty sure `class` is not a type but a keyword... you use class to specify classes, it's not uses (as a type) to declare a variable `class a`... you'd say `NameOfClass a`... so in that usage:

```
class Booger { 
}
```

It would be a keyword.

Reply:
Yeah you're right. I'm still trying to get used to the internal conflicts between the language lawyer and the highlight mindset. I'll fix this.

Reply:
I'm going to move `auto`, `class`, `struct`, `union` and `enum`, as they're all somewhere in the twilight zone of being a keyword/storage specifier and a type specifier. I don't have a strong opinion about this regarding the highlighting class, but language-strictly spoken they're probably more type than keyword. But at least they're considered 'special' for the highlighter -- this is the most important fact.

Reply:
And indeed in templates `class` serves a double role as a typename specifier. With some regular expression magic this could be made a special case, but it's fine like it is.

Reply:
> I'm going to move auto, class, struct, union and enum,

I'd say broadly these are all keywords (from our highlighting POV and historically how we treat them across many grammars).  We don't currently differentiate between "type of storage" vs "type of type"... I think the context of their usage is relevant as well... lets take a made up class-based language:

Next is me just thinking about this seat of my pants:

```js
// define a class
class Person { }

// a new variable of class Person
var bob : class<Person> = Person.new(...)
```

The first `class` is a `keyword`.  You could argue the second `class` is a bit closer to `type` (and that could be worth discussing), but for convenience we'd probably just call it a keyword, rather than try to contextually parse the context.  `Person` of course would be `title.class`.

Reply:
> I don't believe class is a type here? We should probably remove it from the list of types and (in a follow-up PR) add it instead with a `class [IDENT]` multi-matcher rule as done in many other languages - allowing us to also flag the class name as a `title.class`, etc...

Could you point me at a language definition file that uses these multi-matcher rules, or at least a language definition that is a very good example of how a language definition should be done? I'm going to try to study this stuff for the next PR. Thanks!

Reply:
Look at Wren.